### ISSUE TITLE
Improve -s (skip up to date) by checking all dependencies

### ISSUE THREAD

**Issue by sushain97** [unhammer/apertium-get#4]
My use case involves compiling all our pairs (for beta.apertium.org). Precisely:

```
docker-compose exec apy bash -c "cd /source && apertium-get -l | grep -v '^$\|^\s*\#' | sort | xargs -n 1 -P 6 timeout 10m apertium-get"
docker-compose exec apy bash -c "cd /root/apertium-apy && git pull"
docker-compose restart apy html-tools
```

If this were to run every 6 hours, it would take far fewer cycles if it didn't do unnecessary configure/autogen steps for each up-to-date pair.

**Comment by unhammer** [comment]
How do we know that configure/autogen is unnecessary? If installed dependencies are different, those steps might have to set different variables before a compile (and `make` won't run properly to tell us if a compile is necessary without a correct configure). I don't think it should be the default, I want `apertium-get` to be fairly "safe" to rerun for newbies (whether they have uncommitted changes to the language data or their installed libraries changed), but it should be fine as an option `--skip-if-up-to-date` that just checks svn info (… or you could just make a wrapper ;))

By the way, can't your `-P6` lead to races where two pairs have the same monolingual dependencies?

**Comment by sushain97** [comment]
> If installed dependencies are different, those steps might have to set different variables before a compile (and make won't run properly to tell us if a compile is necessary without a correct configure). 

Indeed. From my PR: "This seems somewhat false-negative'y if a corresponding language module is updated but the pair isn't. Any ideas to fix it?"

> but it should be fine as an option --skip-if-up-to-date that just checks svn info

Sounds good! I'll modify my PR.

> By the way, can't your -P6 lead to races where two pairs have the same monolingual dependencies?

Heh, yep.

I've slightly changed it to hopefully prevent that by getting all the monolingual deps first

```
docker-compose exec apy bash -c "cd /source && ./apertium-get -m | grep -v '^$\|^\s*\#' | sort | xargs -n 1 -P 6 timeout 10m ./apertium-get"
docker-compose exec apy bash -c "cd /source && ./apertium-get -l | grep -v '^$\|^\s*\#' | sort | xargs -n 1 -P 6 timeout 10m ./apertium-get"
```

**Comment by unhammer** [comment]
>> If installed dependencies are different, those steps might have to
>> set different variables before a compile (and make won't run
>> properly to tell us if a compile is necessary without a correct
>> configure).
>
> Indeed. From my PR: "This seems somewhat false-negative'y if a
> corresponding language module is updated but the pair isn't. Any ideas
> to fix it?"

The list of deps is created before we even run autogen/configure in pairs (only depends on configure.ac, which is from SVN). We could wait with the `skip_if_up_to_date/is_dep_updated` test until the deps list has been created, then (in `get_pair`) before calling `get_data`, we call `is_dep_updated` on every dep, and if all are updated, we skip that pair. That also means all calls to `is_dep_updated` happen from `get_pair`; I don't think that would be a bad thing.

**Comment by sushain97** [comment]
> We could wait with the skip_if_up_to_date/is_dep_updated test until the deps list has been created, then (in get_pair) before calling get_data, we call is_dep_updated on every dep, and if all are updated, we skip that pair.

Ahh, I like it. Should the current false-negative behavior be merged in and then improved or should it wait for this?

> That also means all calls to is_dep_updated happen from get_pair; I don't think that would be a bad thing.

I think there still needs to be one in `get_data` since installing e.g. `apertium-alt` never touches `get_pair`?

**Comment by unhammer** [comment]
>> We could wait with the skip_if_up_to_date/is_dep_updated test until the deps list has been created, then (in get_pair) before calling get_data, we call is_dep_updated on every dep, and if all are updated, we skip that pair.
>
> Ahh, I like it. Should the current false-negative behavior be merged in and then improved or should it wait for this?

We can merge that first and then improve.

>> That also means all calls to is_dep_updated happen from get_pair; I don't think that would be a bad thing.
>
> I think there still needs to be one in `get_data` since installing e.g. `apertium-alt` never touches `get_pair`?

Oh right, monolinguals can be installed directly … I don't know if there's a simple way to avoid the duplicate check, but perhaps it's fairly fast?
